File: src/lib/voter-segments.ts
import { VoterRecord, VoterSegment, MatchResult, SafeVoterRecord, VoteValue } from '@/types'

const ELECTION_FIELDS = ['VH2024G', 'VH2022G', 'VH2020G', 'VH2024P', 'VH2022P', 'VH2020P'] as const

const VOTED_VALUES = new Set<VoteValue>(['Y', 'A', 'E'])

export function calculateVoteScore(record: VoterRecord | SafeVoterRecord): number {
  let voteCount = 0
  for (const field of ELECTION_FIELDS) {
    if (VOTED_VALUES.has(record[field])) {
      voteCount++
    }
  }
  return voteCount / ELECTION_FIELDS.length
}

export function determineSegment(score: number): VoterSegment {
  if (score >= 0.8) return 'super-voter'
  if (score >= 0.3) return 'sometimes-voter'
  return 'rarely-voter'
}

export interface VoteHistoryDetail {
  election: string
  year: string
  type: 'General' | 'Primary'
  voted: boolean
  method: string
}

export function getVoteHistoryDetail(record: SafeVoterRecord): VoteHistoryDetail[] {
  const elections: { field: typeof ELECTION_FIELDS[number]; year: string; type: 'General' | 'Primary' }[] = [
    { field: 'VH2024G', year: '2024', type: 'General' },
    { field: 'VH2022G', year: '2022', type: 'General' },
    { field: 'VH2020G', year: '2020', type: 'General' },
    { field: 'VH2024P', year: '2024', type: 'Primary' },
    { field: 'VH2022P', year: '2022', type: 'Primary' },
    { field: 'VH2020P', year: '2020', type: 'Primary' },
  ]

  return elections.map(({ field, year, type }) => {
    const value = record[field]
    return {
      election: field,
      year,
      type,
      voted: VOTED_VALUES.has(value),
      method: value === 'Y' ? 'In Person'
        : value === 'A' ? 'Absentee'
          : value === 'E' ? 'Early'
            : 'Did Not Vote',
    }
  })
}

export function segmentResults(results: MatchResult[]) {
  const confirmed = results.filter(r => r.status === 'confirmed' || r.status === 'ambiguous')
  const matched = confirmed.filter(r => r.segment)

  return {
    superVoters: matched.filter(r => r.segment === 'super-voter'),
    sometimesVoters: matched.filter(r => r.segment === 'sometimes-voter'),
    rarelyVoters: matched.filter(r => r.segment === 'rarely-voter'),
    unmatched: results.filter(r => r.status === 'unmatched'),
    totalEntered: results.length,
    totalMatched: matched.length,
  }
}

File: ios/Threshold/Threshold/Utilities/VoterSegments.swift
import Foundation

/// Port of src/lib/voter-segments.ts
enum VoterSegmentCalculator {
    /// Vote codes that count as "voted" — In Person, Absentee, and Early
    private static let votedCodes: Set<String> = ["Y", "A", "E"]

    /// Calculate vote score from 0-100 based on voting history.
    static func calculateVoteScore(voter: SafeVoterRecord) -> Double {
        var score: Double = 0
        let maxScore: Double = 100

        // General elections weighted more heavily
        let generalWeight: Double = 20
        let primaryWeight: Double = 10

        // 2024 General
        if votedCodes.contains(voter.vh2024G) { score += generalWeight }
        // 2022 General
        if votedCodes.contains(voter.vh2022G) { score += generalWeight }
        // 2020 General
        if votedCodes.contains(voter.vh2020G) { score += generalWeight }
        // 2024 Primary
        if votedCodes.contains(voter.vh2024P) { score += primaryWeight }
        // 2022 Primary
        if votedCodes.contains(voter.vh2022P) { score += primaryWeight }
        // 2020 Primary
        if votedCodes.contains(voter.vh2020P) { score += primaryWeight }

        return min(score, maxScore)
    }

    /// Determine voter segment from vote score.
    static func determineSegment(voteScore: Double) -> VoterSegment {
        if voteScore >= 60 {
            return .superVoter
        } else if voteScore >= 30 {
            return .sometimesVoter
        } else {
            return .rarelyVoter
        }
    }

    /// Check if a voter matches the campaign's target universe criteria.
    /// Returns true only if ALL specified criteria match (AND logic).
    static func isInTargetUniverse(voter: SafeVoterRecord, config: TargetUniverseConfig?) -> Bool {
        guard let config else { return false }

        let elections: [(value: String, criterion: String?)] = [
            (voter.vh2024G, config.VH2024G),
            (voter.vh2022G, config.VH2022G),
            (voter.vh2020G, config.VH2020G),
            (voter.vh2024P, config.VH2024P),
            (voter.vh2022P, config.VH2022P),
            (voter.vh2020P, config.VH2020P),
        ]

        var hasCriteria = false
        for (voteValue, criterion) in elections {
            guard let criterion, !criterion.isEmpty else { continue }
            hasCriteria = true

            let voted = votedCodes.contains(voteValue)
            if criterion == "voted" && !voted { return false }
            if criterion == "did-not-vote" && voted { return false }
        }

        return hasCriteria
    }
}
